### PR TITLE
Upload progress recipe

### PR DESCRIPTION
**Pull request by seokjun7410**
Dear. @swankjesse

I improved the `upload progress recipe` discussed in #1471 #1528
It was rewritten based on the discussed requirements.

After discussion, I would like to write a Kotlin recipe
I'm awaiting your comments. 
Thank you for your time

### CODE REVIEW

**Comment by yschimke** [comment]
cc @swankjesse @LouisCAD in case you have thoughts, otherwise I'll review next week

**Comment by seokjun7410** [comment]
If you officially provide the recipe, it will be very helpful to new users

File: samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+		}
+
+		@Override
+		public long contentLength() throws IOException {
+			return requestBody.contentLength();
+		}
+
+		@Override
+		public void writeTo(BufferedSink sink) throws IOException {
+			BufferedSink bufferSink = Okio.buffer(sink(sink));
```
Comment (swankjesse) [inline samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java:106]: supernit: bufferedSink, not bufferSink

File: samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+			requestBody.writeTo(bufferSink);
+			bufferSink.close();
```
Comment (swankjesse) [inline samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java:108]: Remove this, the implementation of `writeTo()` shouldn’t close `sink`

File: samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	public static void main(String... args) throws Exception {
+		new UploadProgress().run();
+	}
+
+	private static class ProgressRequestBody extends RequestBody {
+
+		private final ProgressListener progressListener;
+		private final RequestBody requestBody;
```
Comment (swankjesse) [inline samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java:87]: rename to `delegate` ?

File: samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+				public void write(Buffer source, long byteCount) throws IOException {
+					super.write(source, byteCount);
+					totalBytesWritten += byteCount;
+					progressListener.update(totalBytesWritten, contentLength(), false);
+				}
+
+				@Override
+				public void close() throws IOException {
+					super.close();
+					progressListener.update(totalBytesWritten, contentLength(),true);
```
Comment (swankjesse) [inline samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java:125]: supernit: formatting. Space after `,`

Comment (swankjesse) [reply samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java:125]: MIGHT want to make this only call `progressListener.update()` once even if `close()` is called multiple times. You’re allowed to call `close()` multiple times but listeners shouldn’t be called more than once

**Comment by yschimke** [comment]
@seokjun7410 thanks for the great sample.

**Pull request merged**